### ISSUE TITLE
Dataflow terms and carets

### ISSUE THREAD

**Issue by AndreVanDelft** [scala-subscript/subscript#3]
In Operators.scala, the DataFlowTerms seem deficient:

after the first ~~ arrow there may be multiple +~~ arrows, and then multiple +~/~ arrows.

A DataflowElse is also allowed:

starting with ~/~ and thereafter there may be multiple +~/~ arrows

The parameters in the arrows may have the same syntax as partial function parameters, e.g. in a match statement.
So @ and if should be allowed.

The caret "^" is also not parsed yet, AFAICS.
The caret after brackets should be defined as:

```
[x]^  === ([x])^
```

i.e. a caret on a call to a lambda that performs x.

**Comment by anatoliykmetyuk** [comment]
Also, it would be nice to support a replacement of a `finally` clause - the one that will be executed after the LHS of the dataflow regardless of the throwed exceptions. A special syntax for it would be needed.

**Comment by AndreVanDelft** [comment]
We do not really need special syntax, since this functionality is already covered.
The dataflow operators are defined in terms of do-then and do-else.
In the following, when x succeeds, y is activated;
in case x terminates in failure, z is activated
when y or z succeeds, w is activated.

```
do x then y else z; w
```

In the following forms, there is a missing branch:

```
do x then y; w
do x else z; w
```

These are short hand notations for forms with neutral operands:

```
do x then y  else []; w
do x then [] else  z; w
```

**Comment by anatoliykmetyuk** [comment]
We may have something like:

``` scala
foo ~~(x: T)~~> bar
  +~/~(e: RuntimeException)~~> handle
w
```

Things may go wrong if an exception which is not a subtype of `RuntimeException` is thrown and is unhandled. In this case, we won't advance to `w`, hence its execution is not guaranteed, as in case of `finally`.

**Comment by AndreVanDelft** [comment]
Ok.

```
x ~/~(e:E)~~>y
```

is syntactic sugar for

```
x ~/~> (case (e:E) ==> y)
```

So if in x an exception is thrown of a type that is not a subtype of E then the partial function call will thrown an exception. The whole then fails with the latter exception.

However, it is already possible to construct something like finally:

```
do x else []; y
```

A high level syntax doing the same could well be:

```
do x finally y
```

You may ask why not "try" rather than "do".
Maybe that would be better, and also for try-then-else etc.

**Comment by anatoliykmetyuk** [comment]
Yes, that will do. This means we should always match against "_" in DSL.scala's dataflow definition, so that the executor doesn't fail with MatchError.

**Comment by AndreVanDelft** [comment]
Yes, you are right; in case of a mismatch, the flown exception should end up as the failure result of the whole. Btw would you prefer "try" or "do"? I more and more feel for "try". We could even think of an optional "catch" clause.

**Comment by anatoliykmetyuk** [comment]
Doesn't matter. I think it'd be better to go with `do` - the less of auxiliary syntax the better. Better to teach the users the general case of `do`-`then`-`else` then mimic the usual `try`-`catch`-`finally`. Would make sense to introduce `try` if we had [try with resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) in mind, but since there's no such thing in Scala, it will be ok.

**Comment by anatoliykmetyuk** [comment]
Just [added](https://github.com/scala-subscript/subscript/compare/4387474c8c...a36c28e3c5) the new dataflow. You can see its usage [here](https://github.com/scala-subscript/subscript/blob/master/framework/core-scripts/src/test/scala/subscript/SubScriptSuite.scala). Btw, the whole [core-scripts](https://github.com/scala-subscript/subscript/tree/master/framework/core-scripts) module is now defined to contain core scripts (we can't place them in the core module, since they require macros and hence depend on the core). Currently core scripts are scripts from [Predef](https://github.com/scala-subscript/subscript/blob/master/framework/core-scripts/src/main/scala/subscript/Predef.scala) (core predef [remains](https://github.com/scala-subscript/subscript/blob/master/framework/core/src/main/scala/subscript/CorePredef.scala) in the core to make OperatorsSuite run), the [dataflow](https://github.com/scala-subscript/subscript/blob/master/framework/core-scripts/src/main/scala/subscript/ScriptDSL.scala#L41) nicely defined as a script, and the [object algebra](https://github.com/scala-subscript/subscript/tree/master/framework/core-scripts/src/main/scala/subscript/objectalgebra) module I find highly useful.

What do we do with the carets? What their semantics should be?

**Comment by AndreVanDelft** [comment]
I hope the following is a complete definition.

The following types of terms may have a caret:
1. a code fragment
2. a script call (possibly after implicit conversion)
3. an expression between brackets

If a script body has no terms with carets, and only one term of the form 1 or 2, then that term is implicitly assumed to have a caret.

`[x]^` is shorthand for `([x])^` which is a call to an anonymous script that performs `x`, so form 3 reduces to form 2.

In case the careted term (a script call or a code fragment) succeeds or fails, then the result value of the enclosing script is set to the result value of the careted term.

Note that multiple successes are possible; failure only happens when a term deactivates without having had a recent success.

The result value of a code fragment is its Scala value after successful execution, wrapped in a `Success`, or if it had thrown an exception then it is that exception, wrapped in `Failure` (and the code fragment does not succeed).

The following 3 rules are relevant for dataflow arrows:
1. The result value of `do x then y else z` is the result value of `y` or `z`, whichever is executed.
2. The result value of `do x then y` is the result value of `y`; only after `x` would fail it becomes `x`'s result.
3. The result value of `do x else y` becomes the result value of `x` when that succeeds; only after `y` is activated (i.e. after a failure of x), the result becomes `y`'s result.

### Caret variations

Instead of a single caret, several variations are possible:

#### an upward dataflow arrow followed by a scala lambda

E.g. `x~~^m` `x~~(v:V)~^m`. 
These arrows have a similar syntax, including plusses, as the normal dataflow arrows. `x~~(v:V)~^m` is shorthand for `[x~~(v:V)~>{:m:}]^` etc., except that it may be implemented more efficiently.

#### a double caret `x^^`

The type of the result value (the Success variant) of the enclosing script must be `List[T]` with the result type of `x` assignable to `T`. The enclosing script maintains a working instance of that type which is initially an empty list. When `x` succeeds with pass equal to `p`, then this working list extended to size `p` in case it had been smaller; the values in the list are the default values for `T`. Also the p'th element (with index `p-1`) of the list becomes the result value of `x`, and the result value of the enclosing script becomes a copy of the list (packed in a `Success`). When `x` fails the result value of the enclosing script becomes the failure value of `x` (packed in `Failure`) 

#### a double caret immediately followed by a number

E.g., `x^^1`, `x^^2`, ...
The carets in the enclosing script should each be numbered; all numbers from 1 to the maximum must be present. The result type should be a tuple `(T1,T2, ...)` such that the terms with caret number 1 have a result type that is assignable to T1 etc.
The result value of the enclosing script contains initially such a tuple with the default values for the type. Upon success of a term that has caret `i`, the result value becomes a copy of the old result value, with the i'th element (1-based) replaced by the result value of the succeeding term. 
When such a careted term `x` fails the result value of the enclosing script becomes the failure value of `x` (packed in `Failure`)

### Shorthand notations

`^scalaValue` => `{:scalaValue:}^`
`^scalaValue^` => `{:scalaValue:}^`
`^scalaValue^^` => `{:scalaValue:}^^`
`^scalaValue^^1` => `{:scalaValue:}^^1`
`^scalaValue~~^m` => `{:scalaValue:}~~^m`
etc.

**Comment by AndreVanDelft** [comment]
```
script abs:List[(A,B)] = ..[a^^1 b^^2]^^
```

this could be too hard to infer.
Maybe we should support specification of the type just before the caret, e.g.,

```
script abs:List[(A,B)] = ..[a^^1 b^^2]:(A,B)^^
```

Naively, this then reduces to 

```
script abs:List[(A,B)] = ..([a^^1 b^^2]:Script[(A,B)])^^
```

Normally `[a^^1 b^^2]` would indeed be a `Script[(A,B)]`, but could we specify instead

```
script abs:List[(A,B)] = ..([a^^1 b^^2]:Script[(Any,Any)])^^
```

Probably not, as the type parameter of Script is not in covariant position.
Could it be made covariant?
If not, the reduction should be a bit more complicated, to a lambda Script with the specified type.

Maybe this is unclear; I feel unsecure about the implementation options.

**Comment by anatoliykmetyuk** [comment]
> Probably not, as the type parameter of Script is not in covariant position.
> Could it be made covariant?

Quite possible. We just need to make the result values to be publicly read-only. In the internals, we can set it via a private method that accepts Any, controlling the correctness of the type by some other means:

``` scala
scala> class Foo[+T] {
     |   var _x: Any = _
     |   def x: T = _x.asInstanceOf[T]
     |   def x_=(y: Any) = _x = y
     | }
defined class Foo

scala> new Foo[Int]
res3: Foo[Int] = Foo@3e6fa38a

scala> res3.x
res4: Int = 0

scala> res3.x = 10
res3.x: Int = 10

scala> res3.x
res5: Int = 10
```

**Comment by anatoliykmetyuk** [comment]
Should we also have `^x` capability for local vars? [Use case](https://github.com/scala-subscript/eye-test/blob/master/src/main/scala/eyetest/data/derby/DerbyUserRepo.scala#L21) - this line will become like this:

``` scala
{!result!}^
```

A better version would be `result^`, but since we won't be able to determine whether `result` is a script or a local var at parse time, a disambiguated version may look like `^result`.

**Comment by AndreVanDelft** [comment]
Yes, `^x` should be available for local vars.
The prefix notation is absolutely needed so that there is no confusion with script names, nor with implicitly converted values.

**Comment by anatoliykmetyuk** [comment]
Can be closed, AFAIK.
